### PR TITLE
register sync product scheduled task. closes #10

### LINKED ISSUE #10: Register a Scheduled task to sync products from PayamGostar

### PR DESCRIPTION
**Pull request by MohammadrezaTaghipour**
closes #10 
pull request for asking code review.

note: the unit test code is under developing and will be sent to the repository in a day.

### CODE REVIEW

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        # endregion
+
+        # region Facts
+
+        [Fact]
+        public void GetPgProductMetaDataByIdTest()
+        {
+            // Arrange
+            var target = new PgProductMetaDataService(_pgProductMetaDataRepository.Object);
+            var initialPgProductMetaData = new PgProductMetaData() { Id = 1, Code = "1", ProductId = 1, LastSyncDate = DateTime.UtcNow };
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: in ezafie

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+            // Act
+            var result = target.GetPgProductMetaDataByCode("1");
+
+            // Assert
+            Assert.Equal(initialPgProductMetaData, result);
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: inja codesho moghayese koni kafie

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.Equal(initialPgProductMetaData.Id, result.Id);
+        }
+
+        [Fact]
+        public void GetPgProductMetaDataByCodeTest()
+        {
+            // Arrange
+            var target = new PgProductMetaDataService(_pgProductMetaDataRepository.Object);
+            var initialPgProductMetaData = new PgProductMetaData() { Id = 2, Code = "2", ProductId = 2, LastSyncDate = DateTime.UtcNow };
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: ezafi

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+        [Fact]
+        public void GetPgProductMetaDataByIdTest()
+        {
+            // Arrange
+            var target = new PgProductMetaDataService(_pgProductMetaDataRepository.Object);
+            var initialPgProductMetaData = new PgProductMetaData() { Id = 1, Code = "1", ProductId = 1, LastSyncDate = DateTime.UtcNow };
+
+            // Act
+            var result = target.GetPgProductMetaDataByCode("1");
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: GetPgProductMetaDataById!

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.Equal(initialPgProductMetaData, result);
+            Assert.Equal(initialPgProductMetaData.Code, result.Code);
+        }
+
+        [Fact]
+        public void GetPgProductMetaDataByCodesTest()
+        {
+            // Arrange
+            var target = new PgProductMetaDataService(_pgProductMetaDataRepository.Object);
+            var initialPgProductMetaDataList = new List<PgProductMetaData>
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: in ezafie

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            {
+                new PgProductMetaData() { Id = 1, Code = "1", ProductId = 1, LastSyncDate = DateTime.UtcNow },
+                new PgProductMetaData() { Id = 3, Code = "3", ProductId = 3, LastSyncDate = DateTime.UtcNow },
+            };
+
+            // Act
+            var result = target.GetPgProductMetaDataByCodes(new List<string> { "1", "3" });
+
+            // Assert
+            Assert.NotNull(initialPgProductMetaDataList);
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: Assert rooye counte result va code item haye result begir

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+            Assert.Equal(initialPgProductMetaDataList, result);
+
+            Assert.Equal(initialPgProductMetaDataList.FirstOrDefault().Id, result.FirstOrDefault().Id);
+            Assert.Equal(initialPgProductMetaDataList.FirstOrDefault().Code, result.FirstOrDefault().Code);
+            Assert.Equal(initialPgProductMetaDataList.FirstOrDefault().ProductId, result.FirstOrDefault().ProductId);
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: iteme dovom chi? ;)

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        //[Fact]
+        public void InsertPgProductMetaDataTest()
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: in chera khalie?

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            var initialProduct = target.GetPgProductMetaDataById(1);
+            initialProduct.Code = "10";
+
+            // Act
+            target.UpdatePgProductMetaData(initialProduct);
+
+            var updateProduct = target.GetPgProductMetaDataById(1);
+
+            // Assert
+            Assert.Equal(initialProduct.Code, updateProduct.Code);
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: inja targeto migiri, codesho update mikoni, bad ba code get mikoni bayad code o idish yeki bashe

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+            Assert.Equal(initialPgProductMetaData1.Code, updatepgProductMetaDataList.FirstOrDefault().Code);
+            Assert.Equal(initialPgProductMetaData2.Code, updatepgProductMetaDataList.LastOrDefault().Code);
+
+            Assert.Equal(initialPgProductMetaData1.ProductId, updatepgProductMetaDataList.FirstOrDefault().ProductId);
+            Assert.Equal(initialPgProductMetaData2.ProductId, updatepgProductMetaDataList.LastOrDefault().ProductId);
+
+            Assert.Equal(15, updatepgProductMetaDataList.FirstOrDefault().ProductId);
+            Assert.Equal(20, updatepgProductMetaDataList.LastOrDefault().ProductId);
+        }
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: inam mese ghabli. assert ziad dare.

File: src/Plugin.Tests/Plugin.Tests.csproj [hunk 11]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\xunit.runner.visualstudio.2.1.0\build\net20\xunit.runner.visualstudio.props" Condition="Exists('..\packages\xunit.runner.visualstudio.2.1.0\build\net20\xunit.runner.visualstudio.props')" />
```
Comment (Borzoo) [inline src/Plugin.Tests/Plugin.Tests.csproj]: remove

File: src/Plugin.Tests/packages.config [hunk 12]
```diff
@@ -42,4 +42,5 @@
   <package id="xunit.core" version="2.1.0" targetFramework="net452" />
   <package id="xunit.extensibility.core" version="2.1.0" targetFramework="net452" />
   <package id="xunit.extensibility.execution" version="2.1.0" targetFramework="net452" />
+  <package id="xunit.runner.visualstudio" version="2.1.0" targetFramework="net452" />
```
Comment (Borzoo) [inline src/Plugin.Tests/packages.config]: remove

File: src/Plugin/PayamGostarPluginDependencyRegistrar.cs [hunk 13]
```diff
@@ -23,6 +23,8 @@ public int Order
 
         public void Register(ContainerBuilder builder, ITypeFinder typeFinder, NopConfig config)
         {
+            builder.RegisterGeneric(typeof(Nop.Data.EfRepository<>)).As(typeof(Nop.Core.Data.IRepository<>)).InstancePerLifetimeScope();
```
Comment (Borzoo) [inline src/Plugin/PayamGostarPluginDependencyRegistrar.cs]: in chera injas?

File: src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs [hunk 14]
```diff
@@ -48,11 +49,35 @@ public virtual PgProductMetaData GetPgProductMetaDataByCode(string code)
                         where p.Code == code
                         select p;
 
-            var pgProductMetaData = query.FirstOrDefault();
+            PgProductMetaData pgProductMetaData = null;
+            using (TransactionScope scope = new TransactionScope(TransactionScopeOption.Required, new TransactionOptions { IsolationLevel = System.Transactions.IsolationLevel.ReadUncommitted }))
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs]: chera?

File: src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs [hunk 15]
```diff
@@ -48,11 +49,35 @@ public virtual PgProductMetaData GetPgProductMetaDataByCode(string code)
...
+                pgProductMetaData = query.FirstOrDefault();
+                scope.Complete();
+            }
 
             return pgProductMetaData;
         }
 
+        public virtual IEnumerable<PgProductMetaData> GetPgProductMetaDataByCodes(List<string> codes)
+        {
+            if (codes == null || !codes.Any())
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs]: ino ye extension method koni khoobe, kheili jaha estefade shode. masalan IEnumerable.IsNotNullOrEmpty

File: src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs [hunk 16]
```diff
@@ -48,11 +49,35 @@ public virtual PgProductMetaData GetPgProductMetaDataByCode(string code)
...
+                return null;
+
+            var query = from p in _pgProductMetaDataRepository.Table
+                        where codes.Contains(p.Code)
+                        select p;
+
+            IEnumerable<PgProductMetaData> pgProductMetaDataList = null;
+            using (TransactionScope scope = new TransactionScope(TransactionScopeOption.Required, new TransactionOptions { IsolationLevel = System.Transactions.IsolationLevel.ReadUncommitted }))
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs]: remove

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+        # region Implementation of ITask
+
+        /// <summary>
+        /// Executes a task
+        /// </summary>
+        public virtual void Execute()
+        {
+            try
+            {
+                var productChnages = _productSyncService.GetProductChangesSince(DateTime.Now); // use DateTime.Now for just as a sample
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: code sample too master merge nemikonim

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+
+                if (productChnages.NewProducts != null && productChnages.NewProducts.Any())
+                {
+                    SyncNewProducts(productChnages.NewProducts);
+                }
+
+                if (productChnages.EditedProducts != null && productChnages.EditedProducts.Any())
+                {
+                    SyncEditedProducts(productChnages.EditedProducts);
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: typo productChnages

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+        # region Implementation of ITask
+
+        /// <summary>
+        /// Executes a task
+        /// </summary>
+        public virtual void Execute()
+        {
+            try
+            {
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: inja bayad transaction estefade koni

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+                var pgProductMetaData = _pgProductMetaDataService.GetPgProductMetaDataByCode(p.Code);
+
+                // update product
+                var product = _productService.GetProductById(pgProductMetaData.ProductId);
+                product.Name = p.Name;
+                product.Price = p.UnitPrice;
+                _productService.UpdateProduct(product);
+
+                // update pgProductMetaData
+                pgProductMetaData.Code = p.Code;
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: bala ba code dari getesh mikoni, ezafie

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+
+        protected virtual void InstantiateNewProduct(Septa.PayamGostar.ServiceLayer.Contract.DataContracts.Product sourceProduct, Nop.Core.Domain.Catalog.Product targetProduct)
+        {
+            targetProduct.Name = sourceProduct.Name;
+            targetProduct.Price = sourceProduct.UnitPrice;
+
+            targetProduct.ProductTypeId = (int)ProductType.SimpleProduct;
+            targetProduct.ParentGroupedProductId = 0; // TODO: check whether the product parent code will be received from PayamGostar or not?
+            targetProduct.VisibleIndividually = true;
+            targetProduct.ProductTemplateId = 1;
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: ProductTemplateId too db e? age hast injoori hardcore nakon idisho

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+            }
+        }
+
+        protected virtual void InstantiateNewProduct(Septa.PayamGostar.ServiceLayer.Contract.DataContracts.Product sourceProduct, Nop.Core.Domain.Catalog.Product targetProduct)
+        {
+            targetProduct.Name = sourceProduct.Name;
+            targetProduct.Price = sourceProduct.UnitPrice;
+
+            targetProduct.ProductTypeId = (int)ProductType.SimpleProduct;
+            targetProduct.ParentGroupedProductId = 0; // TODO: check whether the product parent code will be received from PayamGostar or not?
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: ParentGroupedProductId too db e? age hast injoori hardcore nakon idisho

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+            targetProduct.VisibleIndividually = true;
+            targetProduct.ProductTemplateId = 1;
+            targetProduct.VendorId = 0;
```
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: baghie ham hamintor

File: src/Plugin.Tests/App.config [hunk 24]
```diff
@@ -39,6 +39,10 @@
         <assemblyIdentity name="WebGrease" publicKeyToken="31bf3856ad364e35" culture="neutral" />
         <bindingRedirect oldVersion="0.0.0.0-1.6.5135.21930" newVersion="1.6.5135.21930" />
       </dependentAssembly>
+      <dependentAssembly>
+        <assemblyIdentity name="AutoMapper" publicKeyToken="be96cd2c38ef1005" culture="neutral" />
```
Comment (Borzoo) [inline src/Plugin.Tests/App.config]: estefade nashode?

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs (hunk 1 above)
Comment (MohammadrezaTaghipour) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: GetPgProductMetaDataByIdTest()
kolan ezafie?

Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: Initial... Ezafie.

File: src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs (hunk 16 above)
Comment (MohammadrezaTaghipour) [inline src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs]: ?

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs (hunk 22 above)
Comment (MohammadrezaTaghipour) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: az bein field haye Product yeseri ha Requeird hastan.
be una moghe insert bayad ye meghdare default bedam!
va badan kasi ke content site dasteshe mitone default dar biare!
any other idea?

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs (hunk 18 above)
Comment (MohammadrezaTaghipour) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: ?

File: src/Plugin.Tests/App.config (hunk 24 above)
Comment (MohammadrezaTaghipour) [inline src/Plugin.Tests/App.config]: 100 darsad be kar miad!
vali ta inja azash estefade nakardam.

File: src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs (hunk 16 above)
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Catalog/PgProductMetaDataService.cs]: Baraye get chera transaction gozashti?

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs (hunk 18 above)
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: ProductChnages eshtebahe typei dare (typo)

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs (hunk 22 above)
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: Khode nop be onvane default chi mizane?

File: src/Plugin.Tests/PgProductMetaDataServiceTest.cs [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+            Assert.Equal("1", result.FirstOrDefault().Code);
+            Assert.Equal("3", result.LastOrDefault().Code);
+        }
+
+        [Fact]
+        public void InsertPgProductMetaDataTest()
+        {
+            // Arrange
+            var insertedProductId = 1;
+            _pgProductMetaDataRepository.Setup(x => x.Insert(It.IsAny<PgProductMetaData>()))
```
Comment (Borzoo) [inline src/Plugin.Tests/PgProductMetaDataServiceTest.cs]: setup dige nemikhad. hamoon ke call koni o verify koni badesh kafie.

File: src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs (hunk 19 above)
Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: ye transaction balaye methode execute bayad start koni akharesham commit koni

Comment (Borzoo) [inline src/Plugin/PayamGostarService/Tasks/ProductSyncTask.cs]: testaye edit o delete ham ezafe she

**Pull request merged**